README
###############################################
Vimeo Service Module
###############################################

13 May 2013

Maintainer Contact
-----------------------------------------------
Ronald Ferguson (Nickname: r0nn1ef)
<r0nn1ef8580 (at) gmail (dot) com>

Requirements
-----------------------------------------------
SilverStripe 3.0.5
PHP 5.2.x

Documentation
-----------------------------------------------


Installation Instructions
-----------------------------------------------
1. Create a basic account at http://www.vimeo.com/join
2. Log in to your Vimeo user account
3. Register your web application at http://www.vimeo.com/api/applications/new (Ensure that you uncheck the checkbox for "Upload Permissions" as you will not need upload access for this module.)
4. Place this directory in the root of your SilverStripe installation
5. Run /dev/build/
6. The CMS should now have a new "VimeoGallery" page type.
   Create a page and set the parameters correctly in the 'Videos' toggle section to retrieve the videos for display.
7. Add these and visit your public site to reflect the changes


Usage Overview
-----------------------------------------------
The Vimeo module is designed to let you display a gallery of videos from the Vimeo video sharing site. The module also adds a shortcode to embed a Vimeo video
into any text field.

Known issues:
-----------------------------------------------
No know or reported issues.

CHANGELOG
0.1:
- initial release
1.0
- final Silverstripe 2.4 release
2.0
- Initial Silverstripe 3.0
-- updated to Silverstripe 3 framework
-- removed Vimeo API & secret key from _config.php and moved to database config
-- removed prettyPhoto
-- updated shortcode to handle all Vimeo parameters

lang/en_US.php
<?php
// -------------------------------------------------------------------
// This file was automatically generated by the LangBuilder.php script
// Date of creation: 2011/06/27 10:21:13
// -------------------------------------------------------------------
global $lang;


// Output for class or file: VimeoGalleryPage
$lang['en_US']['VimeoGalleryPage']['Settings'] =
	'Vimeo';
$lang['en_US']['VimeoGalleryPage']['SINGULAR_NAME'] =
	'Vimeo Gallery Page';
$lang['en_US']['VimeoGalleryPage']['PLURAL_NAME'] =
	'Vimeo Gallery Pages';
$lang['en_US']['VimeoGalleryPage']['SELECT'] =
	'Select';
$lang['en_US']['VimeoGalleryPage']['USER'] =
	'User';
$lang['en_US']['VimeoGalleryPage']['GROUP'] =
	'Group';
$lang['en_US']['VimeoGalleryPage']['ALBUM'] =
	'Album';
$lang['en_US']['VimeoGalleryPage']['USER_ID_LABEL'] =
	'Vimeo Username/Vimeo Group Name/Vimeo Album ID';
$lang['en_US']['VimeoGalleryPage']['VIDEOS_PER_PAGE'] =
	'Number of videos per page';
$lang['en_US']['VimeoGalleryPage']['SORT_BY'] =
	'Sort by';
$lang['en_US']['VimeoGalleryPage']['NEWEST'] =
	'Newest';
$lang['en_US']['VimeoGalleryPage']['OLDEST'] =
	'Oldest';
$lang['en_US']['VimeoGalleryPage']['MOST_PLAYED'] =
	'Most played';
$lang['en_US']['VimeoGalleryPage']['MOST_COMMENTED'] =
	'Most commented';
$lang['en_US']['VimeoGalleryPage']['MOST_LIKED'] =
	'Most liked';
$lang['en_US']['VimeoGalleryPage']['VIDEO_IN_POPUP'] =
	'Show video in popup?';
$lang['en_US']['VimeoGalleryPage']['POPUP_THEME'] =
	'Popup theme';
$lang['en_US']['VimeoGalleryPage']['DEFAULT'] =
	'Default';
$lang['en_US']['VimeoGalleryPage']['DARK_SQUARE'] =
	'Dark Square';
$lang['en_US']['VimeoGalleryPage']['DARK_ROUNDED'] =
	'Dark Rounded';
$lang['en_US']['VimeoGalleryPage']['LIGHT_SQUARE'] =
	'Light Square';
$lang['en_US']['VimeoGalleryPage']['LIGHT_ROUNDED'] =
	'Light Rounded';
$lang['en_US']['VimeoGalleryPage']['FACEBOOK'] =
	'Facebook';
$lang['en_US']['VimeoGalleryPage']['POPUP_WIDTH'] =
	'Popup width';
$lang['en_US']['VimeoGalleryPage']['POPUP_HEIGHT'] =
	'Popup height';
$lang['en_US']['VimeoGalleryPage']['VIDEO_NOT_FOUND'] =
	'Sorry that video could not be found.';
$lang['en_US']['VimeoGalleryPage']['UNTITLED_VIDEO'] =
	'Untitled Video';


code/VimeoService.php
<?php
/**
 * Service class to interface SilverStripe with the Vimeo video sharing service.
 *
 * @author Ronald Ferguson
 * @version 1.0
 * @package vimeoservice
 */
class VimeoService Extends RestfulService {

	/**
	 * Default player width
	 * @access protected
	 * @static
	 */
	protected static $player_width = 400;

	/**
	 * Default player height
	 * @access protected
	 * @static
	 */
	protected static $player_height = 225;

	/**
	 * API key supplied by Vimeo to use the advanced api
	 * @access protected
	 */
	protected $api_key;

	/**
	 * API key secret supplied by Vimeo for use with the API key
	 * @access protected

	 */
	protected $api_secret_key;

	protected static $token;

	protected static $token_secret;

	/**
	 * Base URL for the Vimeo advanced api
	 * @static
	 */
	public static $api_base_url = 'http://vimeo.com/api/rest/v2';

	/**
	 * @var integer
	 * @access protected
	 */
	protected $video_count;

	/**
	 * @var integer
	 * @access protected
	 */
	protected $page_count;

	/**
	 * @var integer
	 * @access protected
	 */
	protected $pager;

	/**
	 * Constructor
	 * @param Set the cache expiry interva. Defaults to 1 hour (3600 seconds)
	 * @see RestfulService
	 * @return void
	 */
	function __construct($expiry=NULL) {
		parent::__construct(self::$api_base_url, $expiry);
		$this->checkErrors = true;
		$this->setAPIKey(SiteConfig::current_site_config()->VimeoAPIKey);
		$this->setSecretKey(SiteConfig::current_site_config()->VimeoSecretKey);
		$this->pager = FALSE;
	}


	function errorCatch($response){
		$err_msg = $response;
		if(strpos($err_msg, '<') === false) user_error("VimeoService Error: $err_msg", E_USER_ERROR);

		return $response;
	}

	/**
	 * Set the API key supplied by Vimeo
	 * @param string
	 */
	public function setAPIKey($value) {
		$this->api_key = trim($value);
	}

	/**
	 * Set the API key secret provided by Vimeo.
	 * @param string
	 */
	public function setSecretKey($value) {
		$this->api_secret_key = trim($value);
	}

	/**
	 * Returns a collection of Vimeo videos for a given method.
	 * @param string
	 * @param array
	 * @param integer
	 * @return DataObjectSet or boolean false
	 */
	function getVideoFeed($method=NULL, $call_params=array(), $page_limit=10) {
		if(! $this->api_key || ! $this->api_secret_key) {
			user_error('Fatal error: Invalid API keyand/or secret specified!');
			return false;
		}

		// Prepare oauth arguments
        $oauth_params = array(
            'oauth_consumer_key' => $this->api_key,
            'oauth_version' => '1.0',
            'oauth_signature_method' => 'HMAC-SHA1',
            'oauth_timestamp' => time(),
            'oauth_nonce' => $this->_generateNonce()
        );

        $page = $call_params['start'] == 0 ? 1 : (($call_params['start']) / $page_limit + 1);

        if(self::$token) {
        	$oauth_params['oauth_token'] = self::$token;
        }

		$api_params = array(
			'format' => 'php',
			'full_response' => 1,
			'per_page' => $page_limit,
			'page' => $page
		);

        if (!empty($method)) {
            $api_params['method'] = $method;
        }

		// Merge args
        foreach ($call_params as $k => $v) {
            if (strpos($k, 'oauth_') === 0) {
                $oauth_params[$k] = $v;
            }
            else {
                $api_params[$k] = $v;
            }
        }

        // Generate the signature
        $oauth_params['oauth_signature'] = $this->_generateSignature(array_merge($oauth_params, $api_params), self::$api_base_url);

        // Merge all args
        $params = array_merge($oauth_params, $api_params);

		$this->setQueryString($params);

		$response = $this->request();

		$videos = unserialize($response->getBody());




		if(isset($videos->videos)) {
			$results = new ArrayList();
			foreach($videos->videos->video as $video) {
				$results->push(new ArrayData($this->_extractVideoInfo($video)));
			}

			$data = array(
				'videos' => $results,
				'limit' => $page_limit,
				'current_page' => $page,
				'total' => intval($videos->videos->total)
			);

			// since we manually created the dataobjectset, we need to set the pager manually, too.
			$this->setPager($results, $page_limit, $page, intval($videos->videos->total));
		} else {
			$data = array(
				'videos' => FALSE,
				'limit' => $page_limit,
				'current_page' => $page,
				'total' => 0
			);
		}

		return $data;
	}

	/**
	 * Returns a Vimeo video object
	 * @param array
	 * @return ArrayData or boolean false
	 */
	function getSingleVideo($call_params=array()) {
		if(!$this->api_key || ! $this->api_secret_key) {
			user_error('Fatal error: Invalid API keyand/or secret specified!');
			return false;
		}

		// Prepare oauth arguments
        $oauth_params = array(
            'oauth_consumer_key' => $this->api_key,
            'oauth_version' => '1.0',
            'oauth_signature_method' => 'HMAC-SHA1',
            'oauth_timestamp' => time(),
            'oauth_nonce' => $this->_generateNonce()
        );

        if(self::$token) {
        	$oauth_params['oauth_token'] = self::$token;
        }

		$api_params = array(
			'format' => 'php',
			'method' => 'vimeo.videos.getInfo'
		);

		// Merge args
        foreach ($call_params as $k => $v) {
            if (strpos($k, 'oauth_') === 0) {
                $oauth_params[$k] = $v;
            }
            else {
                $api_params[$k] = $v;
            }
        }

        // Generate the signature
        $oauth_params['oauth_signature'] = $this->_generateSignature(array_merge($oauth_params, $api_params), self::$api_base_url);

        // Merge all args
        $params = array_merge($oauth_params, $api_params);

		$this->setQueryString($params);

		$response = $this->request();

		$video = unserialize($response->getBody());

		if(isset($video->video)) {

			foreach($video->video as $video) {
				$data = $this->_extractVideoInfo($video);
				$results =new ArrayData($data);
			}
		} else {
			$results = false;
		}

		return $results;
	}

	/**
	 * @param ArrayList $items
	 * @param int $per_page
	 * @param int $current_page
	 * @param int $total_items
	 */
	protected function setPager(ArrayList $items, $per_page, $current_page, $total_items) {
		$this->pager = new PaginatedList($items, Controller::curr()->getRequest());
		$this->pager->setPageLength($per_page);
		$this->pager->setCurrentPage($current_page);
		$this->pager->setTotalItems($total_items);
		$this->pager->setLimitItems(TRUE);
	}

	public function getPager() {
		return $this->pager;
	}

	/**
	 * Extracts and populates an array structure with the video information returned from Vimeo for a single video object.
	 * @param Object
	 * @return array
	 */
	private function _extractVideoInfo($data) {
		$video = array();
		$video['ID'] = intval($data->id);
		$video['Title'] = new Text(null);
		$video['Title']->setValue(trim($data->title));
		/*
		 * need to clean up the text returned from Vimeo and paragraph format it so we can use
		 * all the nifty SilverStripe text functionality. :)
		 */
		$desc = nl2br(trim($data->description));
		$desc = str_replace("<br />\n<br />", '<br />', $desc);
		$desc = '<p>' . str_replace('<br />', '</p><p>', $desc) . '</p>';
		$video['Description'] = new HTMLText();
		$video['Description']->setValue($desc);

		// video url info
		foreach($data->urls->url as $url) {
			switch($url->type) {
				case 'video':
					$video['Url'] = new Text();
					$video['Url']->setValue(trim($url->_content));
					break;
				case 'mobile':
					$video['MobileUrl'] = new Text();
					$video['MobileUrl']->setValue(trim($url->_content));
					break;
				default:
					// do nothing.
					break;
			}
		}

		// video date info
		$upload_date = trim($data->upload_date);
		// create a SS_Datetime object so we can use familiar functionality.
		$ud = new SS_Datetime();
		$ud->setValue($upload_date);
		$video['UploadDate'] = $ud;
		$modified_date = trim($data->modified_date);
		// create a SS_Datetime object so we can use familiar functionality.
		$md = new SS_Datetime();
		$md->setValue($modified_date);
		$video['ModifiedDate'] = $md;

		// video thumbnail info
		foreach($data->thumbnails->thumbnail as $tn) {
			if(strpos($tn->_content, '_100.jpg') !== false) {
				$video['ThumbSmall'] = new Text();
				$video['ThumbSmall']->setValue(trim($tn->_content));
			} elseif (strpos($tn->_content, '_200.jpg') !== false) {
				$video['ThumbMedium'] = new Text();
				$video['ThumbMedium']->setValue(trim($tn->_content));
			} elseif (strpos($tn->_content, '_640.jpg') !== false) {
				$video['ThumbLarge'] = new Text();
				$video['ThumbLarge']->setValue(trim($tn->_content));
			}
		}

		// User information
		$video['UserName'] = new Varchar(null, 255);
		$video['UserName']->setValue(trim($data->owner->username));
		$video['UserDisplayName'] = new Varchar(null, 255);
		$video['UserDisplayName']->setValue(trim($data->owner->display_name));
		$video['UserRealName'] = new Varchar(null, 255);
		$video['UserRealName']->setValue(trim($data->owner->realname));
		$video['UserID'] = (int)trim($data->owner->id);
		$video['UserUrl'] = new Text();
		$video['UserUrl']->setValue(trim($data->owner->profileurl));
		// User photos
		for($i=0;$i<count($data->owner->portraits->portrait);$i++) {
			$tn = $data->owner->portraits->portrait[$i];
			switch($i) {
				case 0:
					$video['UserPortraitSmall'] = new Text();
					$video['UserPortraitSmall']->setValue(trim($tn->_content));
					break;
				case 1:
					$video['UserPortraitMedium'] = new Text();
					$video['UserPortraitMedium']->setValue(trim($tn->_content));
					break;
				case 2:
					$video['UserPortraitLarge'] = new Text();
					$video['UserPortraitLarge']->setValue(trim($tn->_content));
					break;
				case 3:
					$video['UserPortraitXLarge'] = new Text();
					$video['UserPortraitXLarge']->setValue(trim($tn->_content));
					break;
				default:
					// do nothing.
					break;
			}
		}

		$video['NumberLikes'] = intval(trim($data->number_of_likes));
		$video['NumberPlays'] = intval(trim($data->number_of_plays));
		$video['NumberComments'] = intval(trim($data->number_of_comments));
		$video['Duration'] = intval(trim($data->duration));
		$video['Width'] = intval(trim($data->width));
		$video['Height'] = intval(trim($data->height));
		$video['IsHD'] = new Boolean(null);
		$video['IsHD']->setValue(intval($data->is_hd));
		$video['EmbedPrivacy'] = new Text();
		$video['EmbedPrivacy']->setValue(trim($data->embed_privacy));

		return $video;
	}

	/**
	 * Generates unique identifier used in OAuth signature.
	 * @access protected
	 * @return string
	 */
	protected function _generateNonce() {
		return md5(uniqid(microtime()));
	}

	/**
	 * Generates OAuth signature required to call Vimeo advanced API functions.
	 * @param array
	 * @param string
	 * @param string
	 * @access protected
	 * @return string
	 * @see https://github.com/vimeo/vimeo-php-lib
	 */
	protected function _generateSignature($params, $url, $request_method = 'GET') {
        uksort($params, 'strcmp');
        $params = self::url_encode_rfc3986($params);

        // Make the base string
        $base_parts = array(
            strtoupper($request_method),
            $url,
            urldecode(http_build_query($params, '', '&'))
        );
        $base_parts = self::url_encode_rfc3986($base_parts);
        $base_string = implode('&', $base_parts);

        // Make the key
        $key_parts = array(
            $this->api_secret_key,
            (self::$token_secret) ? self::$token_secret : ''
        );
        $key_parts = self::url_encode_rfc3986($key_parts);
        $key = implode('&', $key_parts);

        // Generate signature
        return base64_encode(hash_hmac('sha1', $base_string, $key, true));
    }

	/**
	 * Gets paged video list for videos of a specific user.
	 * @param string
	 * @param integer
	 * @param integer
	 * @param string
	 * @return DataObjectSet or boolean false
	 * @see getVideoFeed()
	 */
	public function getVideosByUser($userid, $start=0, $per_page=10, $sort=null) {
		$params = array('user_id' => $userid, 'format' => 'php', 'start' => intval($start));
		if($sort) $params['sort'] = $sort;
		$method = 'vimeo.videos.getAll';
		return $this->getVideoFeed($method, $params, $per_page);
	}

	/**
	 * Gets paged video list for videos of a specific Vimeo group.
	 * @param string
	 * @param integer
	 * @param integer
	 * @param string
	 * @return DataObjectSet or boolean false
	 * @see getVideoFeed()
	 */
	public function getVideosByGroup($groupid, $start=0, $per_page=10, $sort=null) {
		$params = array('group_id' => $groupid, 'format' => 'php', 'start' => intval($start));
		if($sort) $params['sort'] = $sort;
		$method = 'vimeo.groups.getVideos';
		return $this->getVideoFeed($method, $params, $per_page);
	}

	/**
	 * Gets paged video list for videos of a specific Vimeo album.
	 * @param integer
	 * @param integer
	 * @param integer
	 * @param string
	 * @return DataObjectSet or boolean false
	 * @see getVideoFeed()
	 */
	public function getVideosByAlbum($albumid, $start=0, $per_page=10, $sort=null) {
		// vimeo.groups.getVideos
		$params = array('album_id' => $albumid, 'format' => 'php', 'start' => intval($start));
		if($sort) $params['sort'] = $sort;
		$method = 'vimeo.albums.getVideos';
		return $this->getVideoFeed($method, $params, $per_page);
	}

	/**
	 * Gets specific Vimeo video.
	 * @param integer
	 * @return ArrayData or boolean false
	 * @see getSingleVideo()
	 */
	public function getVideoById($videoid) {
		$params = array('video_id' => $videoid);
		return $this->getSingleVideo($params);
	}

	/**
	 * URL encode a parameter or array of parameters.
	 * @param mixed
	 * @return mixed
	 */
	public static function url_encode_rfc3986($input) {
        if (is_array($input)) {
            return array_map(array('VimeoService', 'url_encode_rfc3986'), $input);
        }
        else if (is_scalar($input)) {
            return str_replace(array('+', '%7E'), array(' ', '~'), rawurlencode($input));
        }
        else {
            return '';
        }
    }
}

?>

code/VimeoSiteConfig.php
<?php
class VimeoSiteConfig extends DataExtension {
	static $db = array(
		'VimeoAPIKey' => 'Varchar(128)',
		'VimeoSecretKey' => 'Varchar(56)',
		'VimeoDefaultWidth' => 'Int',
		'VimeoDefaultHeight' => 'Int',
		'VimeoPlayerBase' => 'Varchar(128)'
	);

	static $defaults = array(
		'VimeoDefaultWidth' => 640,
		'VimeoDefaultHeight' => 360,
		'VideoPlayerBase' => 'http://player.vimeo.com/video/'
	);

	public function updateCMSFields(FieldList $fields) {
		$fields->addFieldToTab("Root.Main", new HeaderField(null, "Vimeo Settings"));
		$fields->addFieldToTab("Root.Main", new TextField("VimeoAPIKey", "API Key"));
		$fields->addFieldToTab("Root.Main", new TextField("VimeoSecretKey", "Secret Key"));
		$fields->addFieldToTab("Root.Main", new TextField("VimeoDefaultWidth", "Default Video Width"));
		$fields->addFieldToTab("Root.Main", new TextField("VimeoDefaultHeight", "Default Video Height"));
		$fields->addFieldToTab("Root.Main", new TextField("VimeoPlayerBase", "Base URL for video player."));
	}

	public function onBeforeWrite() {
		$this->owner->VimeoDefaultWidth = (int)$this->owner->VimeoDefaultWidth == 0 ? 640 : $this->owner->VimeoDefaultWidth;
		$this->owner->VimeoDefaultHeight = (int)$this->owner->VimeoDefaultHeight == 0 ? 360 : $this->owner->VimeoDefaultHeight;
		$this->owner->VimeoPlayerBase .= (substr($this->owner->VimeoPlayerBase, (strlen($this->owner->VimeoPlayerBase)-1), 1) != '/' ? '/' : '');
	}
}


code/VimeoGalleryPage.php
<?php

class VimeoGalleryPage extends Page {

	public static $singular_name = 'Vimeo Gallery Page';

	public static $plural_name = 'Vimeo Gallery Pages';

	public static $db = array(
		'Method' => 'Int',
		'User' => 'Varchar(100)',
		'VideosPerPage' => 'Int',
		'SortField' => 'Varchar(100)',
		'VideoWidth' => 'Int',
		'VideoHeight' => 'Int',
		'VideoTitle' => 'Int',
		'VideoByLine' => 'Int',
		'VideoPortrait' => 'Int',
		'VideoColor' => 'Varchar(36)',
		'VideoAutoPlay' => 'Int',
		'VideoLoop' => 'Int',
		'VideoAPI' => 'Int',
		'VideoPlayerID' => 'Varchar(128)'
	);

	public static $defaults = array(
		'Method' => 1,
		'ShowVideoInPopup' => TRUE,
		'VideosPerPage' => 10,
		'SortField' => 'UploadDate',
		'VideoWidth' => 640,
		'VideoHeight' => 360,
		'VideoTitle' => 1,
		'VideoByLine' => 1,
		'VideoPortrait' => 1,
		'VideoColor' => '',
		'VideoAutoPlay' => 0,
		'VideoLoop' => 0,
		'VideoAPI' => 1,
		'VideoPlayerID' => ''
	);

	public static $allowed_children = array();

	public static $icon = 'vimeoservice/images/vimeo-file.gif';

	protected $_videos = NULL;

	protected $_current_page = NULL;

	protected $_total_videos = NULL;

	protected $_cachedVideos = NULL;

	function getCMSFields() {
		$fields = parent::getCMSFields();

		$vimeoToggle = ToggleCompositeField::create('Vimeo', "Video Settings",
						array(
							new DropdownField("Method", _t('VimeoGalleryPage.SELECT', "Select"), array(
								'1' => _t('VimeoGalleryPage.USER', 'User'),
								'2' => _t('VimeoGalleryPage.GROUP', 'Group'),
								'3' => _t('VimeoGalleryPage.ALBUM', 'Album')
							)),
							new TextField("User", _t('VimeoGalleryPage.USER_ID_LABEL', "Vimeo Username/Vimeo Group Name/Vimeo Album ID")),
							new DropdownField("VideosPerPage", _t('VimeoGalleryPage.VIDEOS_PER_PAGE', "Number of videos per page"), array(
								'10' => '10',
								'20' => '20',
								'30' => '30',
								'40' => '40',
								'50' => '50'
							)),
							new DropdownField("SortField", _t('VimeoGalleryPage.SORT_BY', "Sort by"), array(
								'newest' => _t('VimeoGalleryPage.NEWEST', 'Newest'),
								'oldest' => _t('VimeoGalleryPage.OLDEST', 'Oldest'),
								'most_played' => _t('VimeoGalleryPage.MOST_PLAYED', 'Most played'),
								'most_commented' => _t('VimeoGalleryPage.MOST_COMMENTED', 'Most commented'),
								'most_liked' => _t('VimeoGalleryPage.MOST_LIKED', 'Most liked')
							)),
							new TextField("VideoWidth", "Video Width"),
							new TextField("VideoHeight", "Video Height"),
							new CheckboxField("VideoTitle", "Display video title on the videos."),
							new CheckboxField("VideoByLine", "Show the user's byline on the videos."),
							new CheckboxField("VideoPortrait", "Show the user's portrait on the videos."),
							new CheckboxField("VideoAutoPlay", "Auto play videos when viewed."),
							new TextField("VideoColor", "Specify the color of the video controls. Make sure that you don't include the #."),
							new CheckboxField("VideoAPI", "Enable Javascript API."),
							new TextField("VideoPlayerID", "Video player ID (required when using the Javascript API.)")
						)
					)->setHeadingLevel(4);

		$fields->insertBefore($vimeoToggle, "Metadata");

		return $fields;
	}

	function VimeoVideos($start = 0) {
		// TODO: Add TRUE SS cache for Vimeo API calls
		// if($this->_cachedVideos) return $this->_cachedVideos;

		$config = SiteConfig::current_site_config();

		$vimeo = new VimeoService();
		$per_page = intval($this->VideosPerPage) < 10 ? 10 : intval($this->VideosPerPage);
		// Used in the pager
		$this->_current_page = $start == 0 ? 1 : (floor($start / $per_page) + 1);

		switch($this->Method) {
			case 1:
				$result = $vimeo->getVideosByUser($this->User, $start, $per_page, $this->SortField);
				break;
			case 2:
				$result = $vimeo->getVideosByGroup($this->User, $start, $per_page, $this->SortField);
				break;
			case 3:
				$result = $vimeo->getVideosByAlbum($this->User, $start, $per_page, $this->SortField);
				break;
			default:
				$result = FALSE;
		}

		if($result) {
			$this->_videos = $result['videos'];
			$this->_total_videos = $result['total'];
		}

		return $this->_videos;
	}

	public function getVideos() {
		return (!$this->_videos ? FALSE : $this->_videos);
	}

	public function getTotalVideos() {
		return (!$this->_total_videos ? 0 : $this->_total_videos);
	}

	public function getCurrentPage() {
		return (!$this->_current_page ? 1 : $this->_current_page);
	}

	function flushCache() {
		parent::flushCache();
		unset($this->_cachedVideos);
	}

	public static function VimeoShortcodeHandler($attributes, $content=NULL, $parser=NULL) {
		if(!isset($attributes['id'])) return '';
		$width = isset($attributes['width']) ? intval($attributes['width']) : SiteConfig::current_site_config()->VimeoDefaultWidth;
		$height = isset($attributes['height']) ? intval($attributes['height']) : SiteConfig::current_site_config()->VimeoDefaultHeight;
		if($width == 0) $width = VimeoService::getDefaultWidth();
		if($height == 0) $width = VimeoService::getDefaultHeight();

		$playerURL = SiteConfig::current_site_config()->VimeoPlayerBase . $attributes['id'];

		$params = array();

		if(isset($attributes['autoplay']) && strtolower($attributes['autoplay']) === 'yes') {
			$params[] = 'autoplay=1';
		}
		if(isset($attributes['color'])) {
			$params[] = 'color=' . str_replace('#', '', trim($attributes['color']));
		}
		if(isset($attributes['title']) && strtolower($attributes['title']) === 'no') {
			$params[] = 'title=0';
		}
		if(isset($attributes['portrait']) && strtolower($attributes['portrait']) === 'no') {
			$params[] = 'portrait=0';
		}
		if(isset($attributes['byline']) && strtolower($attributes['byline']) === 'no') {
			$params[] = 'byline=0';
		}
		if(isset($attributes['loop']) && strtolower($attributes['loop']) === 'yes') {
			$params[] = 'loop=1';
		}
		if(isset($attributes['api']) && strtolower($attributes['api']) == 'yes') {
			$params[] = 'api=1';
			if(isset($attributes['player_id'])) {
				$params[] = 'player_id=' . trim($attributes['player_id']);
			}
		}

		$playerURL .= (!$params ? '' : '?' . implode('&', $params));

		return "<iframe src='{$playerURL}' width='{$width}' height='{$height}' frameborder='0' id='vimeo-video-{$attributes['id']}'></iframe>";
	}

	public function onBeforeWrite() {
		$this->VideoColor = str_replace('#', '', $this->VideoColor);
		parent::onBeforeWrite();
	}

}

class VimeoGalleryPage_Controller extends Page_Controller {

	static $allowed_actions = array(
				'index',
				'view'
			);

	function init() {
		if(Director::fileExists(project() . "/css/VimeoGallery.css")) {
			Requirements::css(project() . "/css/VimeoGallery.css");
		} elseif(Director::fileExists('themes/' . project() . "/css/VimeoGallery.css")) {
			Requirements::css('themes/' . project() . "/css/VimeoGallery.css");
		} else {
			Requirements::css("vimeoservice/css/VimeoGallery.css");
		}

		parent::init();
	}

	function isUserRequest() {
		return $this->Method == 1 ? TRUE : FALSE;
	}

	function isGroupRequest() {
		return $this->Method == 2 ? TRUE : FALSE;
	}

	function getVideo($video_id) {

		if(is_numeric($video_id)) {
			$config = SiteConfig::current_site_config();
			$vimeo = new VimeoService();
			$video = $vimeo->getVideoById($video_id);
			return $video;
		} else {
			return FALSE;
		}
	}

	protected function buildVideoURL() {
		if(!$this->_video) return FALSE;

		$url = 'http://player.vimeo.com/video/' . $this->_video->ID;
		$params = array();

		if($this->VideoAutoPlay) {
			$params[] = 'autoplay=1';
		}
		if($this->VideoColor) {
			$params[] = 'color=' . trim($this->VideoColor);
		}
		if(!$this->VideoTitle) {
			$params[] = 'title=0';
		}
		if(!$this->VideoPortrait) {
			$params[] = 'portrait=0';
		}
		if(!$this->VideoByLine) {
			$params[] = 'byline=0';
		}
		if($this->VideoLoop) {
			$params[] = 'loop=1';
		}
		if($this->VideoAPI) {
			$params[] = 'api=1';
		}
		if($this->VideoPlayerID) {
			$params[] = 'player_id=' . $this->VideoPlayerID;
		}

		return $url . (!$params ? '' : '?' . implode('&', $params));
	}

	public function index() {
		$start = isset($_GET['start']) ? (int)$_GET['start'] : 0;
		return $this->Customise(array('Videos' => $this->VimeoVideos($start)));
	}

	function view() {
		$params = $this->getURLParams();
		$video_id = !$params['ID'] ? '' : $params['ID'];

		if($this->_video = $this->getVideo($video_id)) {
			$data = array('Video' => $this->_video, 'VideoURL' => $this->buildVideoURL());
			return $this->Customise($data);
		} else {
			return $this->httpError(404, _t('VimeoGalleryPage.VIDEO_NOT_FOUND', 'Sorry that video could not be found.'));
		}

	}

	///////////////////////// Page control functions ////////////////////////////

	function Title() {
		$title = $this->Title;

		if($this->_video) {
			$title = $this->_video->Title->getValue() . ' » ' . $title;
		}
		return $title;
	}

	function Breadcrumbs($separator = ' &raquo; ') {

		//Get the default breadcrumbs
        $Breadcrumbs = parent::getBreadcrumbs($separator);
		$Parts = explode($separator, $Breadcrumbs);
		// If we are viewing a single video, add link back to the index action of this controller
		// and add the video title to the breadcrumbs.
		$params = $this->getURLParams();
		if($params['Action'] == 'view') {
			$lastIdx = count($Parts)-1;
			$Parts[$lastIdx] = '<a href="' . $this->Link() . '">' . $Parts[$lastIdx] . '</a>';
			$Parts[] = !$this->_video->Title ? _t('VimeoGalleryPage.UNTITLED_VIDEO', 'Untitled Video') : $this->_video->Title->getValue();
		}

		$Breadcrumbs = implode($separator, $Parts);

		return $Breadcrumbs;
	}

	/**
	 * Custom page control that will override SiteTree:Level($level).
	 *
	 * If we are in the view action and an ID is present, we need to
	 * return TRUE because it's like a child page.
	 *
	 * @param int $level
	 */
	function Level($level) {
		$params = $this->getURLParams();

		return isset($params['ID']) ? 1 : null;
	}

	function PaginatedPages() {
		$pager = new PaginatedList($this->Videos, $this->request);
		$pager->setPageLength($this->VideosPerPage);
		$pager->setCurrentPage($this->CurrentPage);
		$pager->setTotalItems($this->TotalVideos);
		$pager->setLimitItems(TRUE);
		return $pager;
	}
}